src/bootstrap/compile.rs

```diff
@@ -1444,12 +1447,13 @@ pub fn run_cargo(
         };
         for filename in filenames {
             // Skip files like executables
-            if !(filename.ends_with(".rlib")
+            if !((!rlib_only_metadata && filename.ends_with(".rlib"))
                 || filename.ends_with(".lib")
                 || filename.ends_with(".a")
                 || is_debug_info(&filename)
                 || is_dylib(&filename)
-                || (is_check && filename.ends_with(".rmeta")))
+                || ((is_check || rlib_only_metadata) && filename.ends_with(".rmeta"))
+                || filename.contains("jemalloc_sys"))
```

Comment:
Why is jemalloc special-cased?

Reply:
It isn't linked into rustc_driver, but several tools still use it from the sysroot. By the way I think rustc_smir may also need to have an exception.